dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/AnalysisReader.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2016 The Ontario Institute for Cancer Research. All rights reserved.                             
+ *                                                                                                               
+ * This program and the accompanying materials are made available under the terms of the GNU Public License v3.0.
+ * You should have received a copy of the GNU General Public License along with                                  
+ * this program. If not, see <http://www.gnu.org/licenses/>.                                                     
+ *                                                                                                               
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY                           
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES                          
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT                           
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,                                
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED                          
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;                               
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER                              
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN                         
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.icgc.dcc.imports.gene.reader;
+
+import static org.icgc.dcc.imports.gene.core.Sources.ANALYSIS_URI;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+public class AnalysisReader {
+
+  private final Map<String, String> analysisMap = new HashMap<>();
+
+  public AnalysisReader read() {
+    BaseReader.read(ANALYSIS_URI, line -> {
+      String id = line[0];
+      String gffSource = line[6];
+
+      // We can support additional programs/algorithms for protein domains here
+      if ("pfam".equals(gffSource)) {
+        analysisMap.put(id, gffSource);
+      }
+    });
+    return this;
+  }
+
+}
```

Comment:
Consider this alternative design for the readers:

# Base Class

```java
@Slf4j
@RequiredArgsConstructor
public abstract class TsvReader {

  /**
   * Configuration
   */
  @NonNull
  private final String uri;

  @SneakyThrows
  protected Stream<List<String>> readRecords() {
    log.info("Reading '{}'...", uri);

    return reader().lines().map(TAB::splitToList);
  }

  private BufferedReader reader() throws MalformedURLException, IOException {
    val gzip = new GZIPInputStream(new URL(uri).openStream());
    val reader = new InputStreamReader(gzip, UTF_8);

    return new BufferedReader(reader);
  }

}
```
# Example Reader
```java
public class AnalysisReader2 extends TsvReader {

  public AnalysisReader2(String uri) {
    super(uri);
  }

  public Map<String, String> read() {
    return readRecords()
        .filter(this::isPfam)
        .collect(toMap(this::getId, this::getGffSource));
  }

  private boolean isPfam(List<String> record) {
    return "pfam".equals(getGffSource(record));
  }

  private String getId(List<String> record) {
    return record.get(0);
  }

  private String getGffSource(List<String> record) {
    return record.get(6);
  }

}
```

This is a more functional design that leverages the Java 8 streams API. Parsing is handled to the base class. The `Stream<List<String>>` abstraction allows powerful streaming capabilities, including filtering, mapping and collecting. You could take this a step further and create other base classes that handle certain data types (Map<String String>) and only require a developer to specify a key / value mapper. In the end, this design gives the implementer flexibility and statelessness, with a minimal amount of code.